### ISSUE TITLE
Error opening jacocoagent.jar due to long path

### ISSUE THREAD

**Issue by brianeray** [gradle/gradle#949]
Adapted from [GRADLE-3547](https://issues.gradle.org/browse/GRADLE-3547) entered by @wolfs, in turn from this [Gradle thread](https://discuss.gradle.org/t/error-opening-jacocoagent-jar-due-to-long-path-to/19250) from Vital Lobachevskij.

### Expected Behavior
Using the [jacoco](https://docs.gradle.org/current/userguide/jacoco_plugin.html) plugin should never yield an `Error opening zip file or JAR manifest missing` error regardless of the length of the path to the project (build) directory and to the jacocoagent.jar nested inside of it.

### Current Behavior
In a project whose build directory path is so long as to nest the expanded jacoco agent JAR such that the full path is somewhere beyond ~160 or ~170 characters, `gradle test` errors with the message beginning like the following.

```
Error opening zip file or JAR manifest missing : ../../build/components/my.long.subproject.name/tmp/expandedArchives/org.jacoco.agent-0.7.6.201602180812.jar_29dabwwwl4sb8maj0h13tor2d/jacocoagent.jar [...]
```

### Context
In my use case I want to run code coverage on Jenkins agents whose workspace (and thus Gradle build) directories are nested deeply enough to almost always push the jacoco agent's full path past the 160-180 threshhold that triggers the above error.

### Steps to Reproduce (for bugs)
**Note:** Though the OP described this as stemming from a long subproject name, I suspect the error would occur any time the file path length is somewhere beyond 160 or 170--at least on Windows. Here are the original steps:

When ‘jacoco’ plugin is applied for all subproject in my multiproject build

```
subprojects {
  apply plugin: 'jacoco'
}
```

gradle test failed with the error like this one:

```
:my.long.subproject.name:test
Error opening zip file or JAR manifest missing : ../../build/components/my.long.subproject.name/tmp/expandedArchives/org.jacoco.agent-0.7.6.201602180812.jar_29dabwwwl4sb8maj0h13tor2d/jacocoagent.jar
Error occurred during initialization of VM
agent library failed to init: instrument
Could not write standard input into: Gradle Test Executor 2.
java.io.IOException: The pipe is being closed
        at java.io.FileOutputStream.writeBytes(Native Method)
        at java.io.FileOutputStream.write(FileOutputStream.java:326)
        at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
        at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
        at org.gradle.process.internal.streams.ExecOutputHandleRunner.run(ExecOutputHandleRunner.java:56)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
:my.long.subproject.name:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':my.long.subproject.name:test'.
> Process Gradle Test Executor 2 finished with non-zero exit value 1
```

I've found that Gradle starts 'Gradle Test Executor' processes using commands like this one:

```
java.exe -Djava.security.manager=worker.org.gradle.process.internal.worker.child.BootstrapSecurityManager -javaagent:../../build/components/my.long.subproject.name/tmp/expandedArchives/org.jacoco.agent-0.7.6.201602180812.jar_29dabwwwl4sb8maj0h13tor2d/jacocoagent.jar=destfile=../../build/components/my.long.subproject.name/jacoco/test.exec,append=true,inclnolocationclasses=false,dumponexit=true,output=file,jmx=false -Xms256m -Xmx1024m -Dfile.encoding=windows-1251 -Duser.country=US -Duser.language=en -Duser.variant -ea -cp C:\Users\Vital_Lobachevskij\.gradle\caches\2.14\workerMain\gradle-worker.jar worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Test Executor 2'
```

It works in the same way for all subprojects. But, if the subproject name is rather long, e.g. ‘my.long.subproject.name’, Gradle can’t open jacocoagent.jar. On my local, full path to this jar for ‘my.long.subproject.name’ is
`d:\Projects\...\build\components\my.long.subproject.name\tmp\expandedArchives\org.jacoco.agent-0.7.6.201602180812.jar_29dabwwwl4sb8maj0h13tor2d\jacocoagent.jar` (179 chars) and the error is thrown.

If I move project folder to the root of disk D, the full path will be shorter (156 chars) and unit tests are passed successfully and code coverage is generated.

### Your Environment
OP's env: Gradle 2.14 on Windows 10
My env: Gradle 3.1 and Zulu OpenJDK 1.8.0_102-b14 on Windows 7 Pro

**Comment by vzhemevko** [comment]
I had the same issue, can confirm that reducing full path length fixes the issue.

**Comment by BalzGuenat** [comment]
Is there a workaround for this if the project path or name cannot be changed? Are there any plans for fixing this (if even possible in Gradle, as opposed to JaCoCo)?

**Comment by BalzGuenat** [comment]
Oh and just by the way, this affects Gradle version 5.6.1 still.

**Comment by anupamgangrade** [comment]
We are facing an issue on similar lines where commands like '_gradle build_' and '_gradle test'_ runs successfully but if I run '_gradle clean build_' it fails with below error:

```
> Task :compileTestJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Could not write standard input to Gradle Test Executor 2.
java.io.IOException: The pipe is being closed
        at java.io.FileOutputStream.writeBytes(Native Method)
        at java.io.FileOutputStream.write(FileOutputStream.java:326)
        at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
        at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
        at org.gradle.process.internal.streams.ExecOutputHandleRunner.forwardContent(ExecOutputHandleRunner.java:66)
        at org.gradle.process.internal.streams.ExecOutputHandleRunner.run(ExecOutputHandleRunner.java:51)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:748)

> Task :test
Error occurred during initialization of VM
agent library failed to init: instrument
Error opening zip file or JAR manifest missing : build/tmp/expandedArchives/org.jacoco.agent-0.8.4.jar_7bfc4fd986d61dbf92065775ff3714a0/jacocoagent.jar
Error occurred during initialization of VM
agent library failed to init: instrument
Error opening zip file or JAR manifest missing : build/tmp/expandedArchives/org.jacoco.agent-0.8.4.jar_7bfc4fd986d61dbf92065775ff3714a0/jacocoagent.jar

```
You can see that path is not that big. What I have noticed is during '_gradle build_' and '_gradle test_' it is able to create '_expandedArchives_' folder under '_build/tmp_' location and hence tests run successfully but while running 'gradle clean build' it cleans build folder but can not create the required folder structure again.

Any leads on this issue. How can I resolve this.

Here is the build environment:
Gradle - 5.6.2
Jacoco - 0.8.4
IntelliJ - 2019.2.3 Ultimate Edition
OS - Windows 7

**Comment by crockeea** [comment]
I submit that a system that auto-closes tickets because no one on your end looked at it is sub-optimal. It doesn't make these issues go away.

**Comment by Adil93** [comment]
I am facing the same issue, Any known fix/workaround for this?

**Comment by Hailong-am** [comment]
> We are facing an issue on similar lines where commands like '_gradle build_' and '_gradle test'_ runs successfully but if I run '_gradle clean build_' it fails with below error:
> 
> ```
> > Task :compileTestJava
> Note: Some input files use or override a deprecated API.
> Note: Recompile with -Xlint:deprecation for details.
> Note: Some input files use unchecked or unsafe operations.
> Note: Recompile with -Xlint:unchecked for details.
> Could not write standard input to Gradle Test Executor 2.
> java.io.IOException: The pipe is being closed
>         at java.io.FileOutputStream.writeBytes(Native Method)
>         at java.io.FileOutputStream.write(FileOutputStream.java:326)
>         at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
>         at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
>         at org.gradle.process.internal.streams.ExecOutputHandleRunner.forwardContent(ExecOutputHandleRunner.java:66)
>         at org.gradle.process.internal.streams.ExecOutputHandleRunner.run(ExecOutputHandleRunner.java:51)
>         at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
>         at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
>         at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
>         at java.lang.Thread.run(Thread.java:748)
> 
> > Task :test
> Error occurred during initialization of VM
> agent library failed to init: instrument
> Error opening zip file or JAR manifest missing : build/tmp/expandedArchives/org.jacoco.agent-0.8.4.jar_7bfc4fd986d61dbf92065775ff3714a0/jacocoagent.jar
> Error occurred during initialization of VM
> agent library failed to init: instrument
> Error opening zip file or JAR manifest missing : build/tmp/expandedArchives/org.jacoco.agent-0.8.4.jar_7bfc4fd986d61dbf92065775ff3714a0/jacocoagent.jar
> ```
> 
> You can see that path is not that big. What I have noticed is during '_gradle build_' and '_gradle test_' it is able to create '_expandedArchives_' folder under '_build/tmp_' location and hence tests run successfully but while running 'gradle clean build' it cleans build folder but can not create the required folder structure again.
> 
> Any leads on this issue. How can I resolve this.
> 
> Here is the build environment: Gradle - 5.6.2 Jacoco - 0.8.4 Java - 8 IntelliJ - 2019.2.3 Ultimate Edition OS - Windows 7

Have same issue, Gradle 8.1.1 JDK 17

**Comment by Fornori** [comment]
We have the same isisue gradle 7.5.1 jacoco 0.8.8 MacOs

**Comment by edudar** [comment]
`clean build` started failing for me out of the blue, even for projects/configurations that worked before. Gradle 8.1, 8.5, 8.6, 8.7, does not matter. As mentioned above, simply `build` or `check` works. 

While inspecting DEBUG logs and trying to debug what Gradle does internally, I noticed that file resolution for jacocoAgent and lock acquisition on `/build/tmp/.cache/expanded` happens before tasks start. As a result, Gradle properly unpacks jacoco JAR and gets a valid path, but `:clean` task just wipes the whole `build` dir after that.

**Comment by ilkruzn** [comment]
I'm having the same issue on a Mac M3 with IDEA 2024.1 update. `clean build` fails but just `build` is fine.

**Comment by JAkutenshi** [comment]
Up the issue and approve @edudar 's  assumption: idea-IU-233.15026.9 (previous build from now) works fine, current build idea-IU-241.14494.240 has the issue. My issue was like that with `jacocoagent.jar`, after downgrading build is fine again.
![image](https://github.com/gradle/gradle/assets/7635910/9eef6abf-d3bb-47d3-b9e3-81ba3a5a08bb)

**Comment by juheba** [comment]
Im facing the same issue:
* IntelliJ IDEA 2024.1 (Ultimate Edition), Build #IU-241.14494.240, built on March 28, 2024
* gradle-7.6.1
* useing plugin { id 'jacoco' } without a specific version

![image](https://github.com/gradle/gradle/assets/47452856/c6bca4ba-a6ee-46b6-ad0c-f725bc5b0b49)

**Comment by jbartok** [comment]
Sorry for the late reply.
On the Gradle side, we have made changes to long path support since the Gradle versions mentioned above. 
Also, there are related fixes in IntelliJ 2024.1.1.
We are considering this issue obsolete. Please open a new one if the problem still persists.
